docs/collaborative-pinsets-rfc.md

```diff
@@ -0,0 +1,131 @@
+
+
+### Dealing with malicious actors
+
+There are several ways that a malicious peer might try to interfere with the activity of a collaborative cluster. In general, we should aim to have a working cluster when a majority of the peers in it are not malicious.
+
+* Having a **trusted peerset** makes it an easy target for DDoS attacks on the swarm.
+* A bad peer can pretend to be pinning something but not do it
+* A bad peer can pretend to have infinite space and be always allocated things
+
+Many of these problems (and others) have been (or will be) addressed by Filecoin. It is not IPFS Cluster goal to overlap with Filecoin in this feature, but rather to offer a simple way to "follow" a pinset which has less overhead and requirements than the current Raft consensus (in the future it will be good to study how Filecoin can be used as the consensus layer for cluster). For the moment:
+
+* We will recommend high replication factors which make it less likely that all allocations for a Cid belong to malicious actors
+* We will have to change the allocator strategy to something that randomly select a subset of peers first, and then applies the strategy. This should as well make it less likely for malicious peers to always appear in front of the allocation candidates.
+* We will have to adapt the actual replication factor dynamically: as new peers join the cluster, we'd want to allocate content to them. It helps thinking in percentages: i.e. a CID should be allocated to 10% of the peers. As peers grow, new peers should start tracking that content. Open questions around allocation strategies are discussed in the next section.
+
+### Allocations: open problems
+
+Currently the allocations for a Cid are part of the shared state and consist of a list of peer IDs, decided when the item was added.
+
+In a scenario where peers come and go and come back, this strategy feels suboptimal (although it would work on principle). We should probably work on an allocator component which can efficiently track and handle allocations as peers come and go. For example, if the minimum allocation factor cannot be reached, cluster should still pin the items and track them as underpinned and, as new peers join, it should allocate underpinned items to them. As peers go away, cluster should efficiently identify which pins need re-allocation.
+
+Perhaps the whole allocation format should be re-thought, allowing each of the peers to detect items which have not reached a high-enough replication factor and tracking them. Then these peers would inform the allocator that they are going to worry about the items. This is, again, a difficult problem which Filecoin has solved properly by providing a market where peers can offer to store content.
```

Comment:
I think a useful differentiator between this project and filecoin is the heightened trust among participants.  It's probably best to focus on uses that make use of higher trust assumptions to avoid FC overlap.